### PR TITLE
Fixes Bread Smite Causing Some Fucked Up Shit

### LINKED ISSUE #69809: You can pick yourself up as bread, runechat moves to your hand
## Reproduction:
Smite a player into bread. The player can pick themselves up with the LEFT hand only. Their runechat speech also floats from the bread in their own hand, in addition to their proper location.

image

### PR DESCRIPTION
**Pull request by san7890**
## About The Pull Request

Hey there,

So basically, when you had the bread smite done on you, you were _just_ added to the contents of the bread. Nothing more. That means that you could pick it up. You couldn't add it to your bag (it would always return back into your hand(?)), but it would create some weird oddities that was just cursed in general. Let's make it so you can't hold the container that you are contained within by giving you HANDS_BLOCKED (if you're a mob since i don't wanna bother with trait buildup bullshit, because apparently you can call Breadify on all atom movables i believe?).

## Why It's Good For The Game

Fixes #69809.

This was just fucked up.

## Changelog

:cl:
fix: When you are smited into an enbreaded tomb, you can no longer pick up said tomb.
/:cl:

### CODE REVIEW

**Review by MrMelbert — COMMENTED** [review]
Also it's worth considering that I've seen quite a few times admins ask, "how do you unbread someone?", and right now the answer is simply "remove from contents"

This makes it a bit more complicated to truly unbread someone (you have to futz around with traits). It's worth noting, not sure I have an alternative suggestion though

**Comment by san7890** [comment]
> Also it's worth considering that I've seen quite a few times admins ask, "how do you unbread someone?", and right now the answer is simply "remove from contents"
> 
> This makes it a bit more complicated to truly unbread someone (you have to futz around with traits). It's worth noting, not sure I have an alternative suggestion though

In an earlier iteration of this PR I thought about adding an "unbreadable" version but... blargh. I wonder if you could do an `unbreadify` proc, but that's even more work.

File: code/modules/admin/verbs/adminfun.dm [hunk 1]
```diff
@@ -217,9 +217,13 @@
 		return
 	smite.effect(src, target)
 
-/proc/breadify(atom/movable/target)
+///"Turns" people into bread. Really, we just add them to the contents of the bread food item.
+/proc/breadify(atom/movable/target, smite = FALSE)
 	var/obj/item/food/bread/plain/bread = new(get_turf(target))
 	target.forceMove(bread)
+	if(!ismob(target)) // sanity check
+		return
```
[the remark this answers is not in the document]
Comment (san7890) [reply code/modules/admin/verbs/adminfun.dm]: ca61233daee78bd933ffafd37bba6dc23b488c6e

**Comment by zxaber** [comment]
You could register a signal from the mob on the bread object to watch for atom_moved, and then remove the trait if `loc != src` (and maybe destroy src after).

**Comment by AnturK** [comment]
Just make bread subtype (or just set up the signals at smite time) that incapacitates anyone you put in there and removes it on exit. Ideally it would be an element/component.

**Comment by san7890** [comment]
> Though i'm curious what changed because i specifically tested this when adding the smite.

Knee-jerk response would be storage datumization?

**Comment by san7890** [comment]
this is my first time working with components so let me know if it's jank

File: code/datums/components/itembound.dm [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	/// The container that the subject is inside of
+	var/atom/movable/container
+	/// Detect any movement of the container
+	var/datum/movement_detector/move_tracker
+
+/datum/component/itembound/Initialize(tomb)
+	if(!ismovable(parent))
+		return COMPONENT_INCOMPATIBLE
+	container = tomb
+	move_tracker = new(parent,CALLBACK(src,.proc/verify_containment))
```
Comment (SuperNovaa41) [inline code/datums/components/itembound.dm:12]:
```suggestion
	move_tracker = new(parent, CALLBACK(src, .proc/verify_containment))
```
simple style stuff

File: code/datums/components/itembound.dm [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+/// Ensure that when we move, we still are in the container. If not in the container, remove all the signals.
+/datum/component/itembound/proc/verify_containment()
+	if(container.contains(parent))
+		return
+	REMOVE_TRAIT(parent, TRAIT_INCAPACITATED, SMITE_TRAIT)
+	qdel(src)
+
+/datum/component/itembound/Destroy(force, silent)
+	QDEL_NULL(move_tracker)
+	. = ..()
```
Comment (SuperNovaa41) [inline code/datums/components/itembound.dm:25]:
```suggestion
	return ..()
```

File: code/datums/components/itembound.dm [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+/datum/component/itembound/Initialize(tomb)
+	if(!ismovable(parent))
+		return COMPONENT_INCOMPATIBLE
+	container = tomb
+	move_tracker = new(parent,CALLBACK(src,.proc/verify_containment))
+
+	ADD_TRAIT(parent, TRAIT_INCAPACITATED, SMITE_TRAIT)
+
+/// Ensure that when we move, we still are in the container. If not in the container, remove all the signals.
```
Comment (SuperNovaa41) [inline code/datums/components/itembound.dm:16]:
```suggestion
/// Ensure that when we move, we still are in the container. If not in the container, remove all the traits.
```
nit picky but you're not actually using signals here?

File: code/datums/components/itembound.dm [hunk 5]
```diff
@@ -0,0 +1,26 @@
+/// When a movable has this component AND they are in the contents of a container, they will no longer be able to use their hands and be immobilized until they are removed from the container. So far, this is only useful for smites.
+/datum/component/itembound
+	/// The container that the subject is inside of
+	var/atom/movable/container
+	/// Detect any movement of the container
+	var/datum/movement_detector/move_tracker
+
+/datum/component/itembound/Initialize(tomb)
```
Comment (SuperNovaa41) [inline code/datums/components/itembound.dm:8]:
```suggestion
/datum/component/itembound/Initialize(container)
```

File: code/datums/components/itembound.dm [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+	if(!ismovable(parent))
+		return COMPONENT_INCOMPATIBLE
+	container = tomb
```
Comment (SuperNovaa41) [inline code/datums/components/itembound.dm:11]:
```suggestion
	src.container = container
```

**Review by SuperNovaa41 — CHANGES_REQUESTED** [review]

**Comment by SuperNovaa41** [comment]
Looks like it should be fine other than style stuff

File: code/datums/components/itembound.dm [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+	move_tracker = new(parent,CALLBACK(src,.proc/verify_containment))
+
+	ADD_TRAIT(parent, TRAIT_INCAPACITATED, SMITE_TRAIT)
+
+/// Ensure that when we move, we still are in the container. If not in the container, remove all the signals.
+/datum/component/itembound/proc/verify_containment()
+	if(container.contains(parent))
```
Comment (SuperNovaa41) [inline code/datums/components/itembound.dm:18]:
```suggestion
	if(container.contains(parent) || !QDELETED(container))
```
also should probably check to see if the container still actually exists

File: code/datums/components/itembound.dm (hunk 4 above)
Comment (san7890) [reply code/datums/components/itembound.dm:16]: cruft from earlier implementation

File: code/datums/components/itembound.dm (hunk 7 above)
Comment (san7890) [reply code/datums/components/itembound.dm]: Actually, is this necessary? If the container is QDEL'd, then all of the contents within (including the movable that's holding the component) should also be deleted, right? Is there some edge case I might be missing?

**Review by MrMelbert — CHANGES_REQUESTED** [review]
local man accidentally re-invents soulstone component
image

File: code/datums/components/itembound.dm [hunk 8]
```diff
@@ -0,0 +1,26 @@
+/// When a movable has this component AND they are in the contents of a container, they will no longer be able to use their hands and be immobilized until they are removed from the container. So far, this is only useful for smites.
+/datum/component/itembound
+	/// The container that the subject is inside of
+	var/atom/movable/container
```
Comment (JohnFulpWillard) [inline code/datums/components/itembound.dm:4]: Would it be better if this was a weakref?

Comment (san7890) [reply code/datums/components/itembound.dm:4]: how do i weakref

**Comment by san7890** [comment]
> local man accidentally re-invents soulstone component image

my behavior is genericized which makes it better i think maybe maybe not

File: code/datums/components/itembound.dm (hunk 8 above)
Comment (SuperNovaa41) [reply code/datums/components/itembound.dm:4]: yes it would

Comment (san7890) [reply code/datums/components/itembound.dm]: okay

**Pull request merged**